src/PowerShell/PowerShellManagerPool.cs

```diff
@@ -15,37 +20,80 @@ namespace Microsoft.Azure.Functions.PowerShellWorker.PowerShell
     /// </summary>
     internal class PowerShellManagerPool
     {
-        private readonly ILogger _logger;
-        // Today we don't really support the in-proc concurrency. We just hold an instance of PowerShellManager in this field.
-        private PowerShellManager _psManager;
+        private readonly int _upperBound;
+        private readonly MessagingStream _msgStream;
+        private readonly BlockingCollection<PowerShellManager> _pool;
+        private int _poolSize;
 
         /// <summary>
         /// Constructor of the pool.
         /// </summary>
-        internal PowerShellManagerPool(ILogger logger)
+        internal PowerShellManagerPool(MessagingStream msgStream)
         {
-            _logger = logger;
+            string upperBound = Environment.GetEnvironmentVariable("PSWorkerInProcConcurrencyUpperBound");
+            if (string.IsNullOrEmpty(upperBound) || !int.TryParse(upperBound, out _upperBound))
+            {
+                _upperBound = 1;
+            }
+
+            _msgStream = msgStream;
+            _pool = new BlockingCollection<PowerShellManager>(_upperBound);
         }
 
         /// <summary>
         /// Initialize the pool and populate it with PowerShellManager instances.
-        /// When it's time to really implement this pool, we probably should instantiate PowerShellManager instances in a lazy way.
-        /// Maybe start from size 1 and increase the number of workers as needed.
+        /// We instantiate PowerShellManager instances in a lazy way, starting from size 1 and increase the number of workers as needed.
         /// </summary>
-        internal void Initialize()
+        internal void Initialize(string requestId)
         {
-            _psManager = new PowerShellManager(_logger);
+            var logger = new RpcLogger(_msgStream);
+
+            try
+            {
+                logger.SetContext(requestId, invocationId: null);
+                _pool.Add(new PowerShellManager(logger));
+                _poolSize = 1;
+            }
+            finally
+            {
+                logger.ResetContext();

```

Comment:
don't think my question in the last PR was addressed - why do we need to ResetContext here?

Reply:
Actually I'm not seeing why we need to pass in requestId to this function.

Reply:
Oh, sorry that I missed your comments there.
We set the context here because logging code may be hit during the initialization when processing `profile.ps1`. But the context doesn't make sense anymore after the initialization. When processing a invocation request, we will set up the context again before returning the checked-out `PowerShellManager`. Therefore, we reset the context here (after finishing creating the first `PowerShellManager`) to make it clean.

Reply:
> Actually I'm not seeing why we need to pass in requestId to this function.

When there is error from invoking the profile, we write out user log, which requires the `RequestId`. Yes, for know, it seems `RequestId` is just an empty string for all requests ... but we need to play safe.